File: src/action_hello_world/scripts/action_hello_world_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# 1. 导入依赖包
import rospy
import actionlib
from action_hello_world.msg import FindPrimesAction, FindPrimesGoal, FindPrimesResult, FindPrimesFeedback

def done_cb(state, result):
    """处理最终结果回调"""
    if state == actionlib.GoalStatus.SUCCEEDED:
        primes_str = " ".join(str(n) for n in result.primes)
        rospy.loginfo(f"{result.number} 以内的质数有: [ {primes_str} ]")

def active_cb():
    """目标激活回调"""
    rospy.loginfo("开始查找...")

def feedback_cb(feedback):
    """处理反馈回调"""
    status_str = "是" if feedback.is_prime else "不是"
    rospy.loginfo(f"当前数字: {feedback.number}, {status_str}质数")

if __name__ == "__main__":
    # 2. 初始化ROS节点
    rospy.init_node("action_hello_world_client")
    
    # 3. 实例化Action客户端对象
    client = actionlib.SimpleActionClient("/find_primes", FindPrimesAction)
    
    # 4. 等待服务端启动
    rospy.loginfo("等待服务端启动...")
    client.wait_for_server()
    rospy.loginfo(f"添加任务前的状态: {client.get_state()}")
    
    # 5. 实例化目标对象
    goal = FindPrimesGoal()
    goal.number = 12
    rospy.loginfo(f"查找 {goal.number} 以内的质数")
    
    # 6. 发送目标任务
    client.send_goal(goal, done_cb=done_cb, active_cb=active_cb, feedback_cb=feedback_cb)
    rospy.loginfo(f"添加任务后的状态: {client.get_state()}")
    
    rospy.sleep(1.0)
    rospy.loginfo(f"执行任务时的状态: {client.get_state()}")
    
    # 7. 等待任务完成
    client.wait_for_result(rospy.Duration(1000.0))  # 1000秒超时
    rospy.loginfo(f"任务执行完的状态: {client.get_state()}")

File: src/action_hello_world/src/action_hello_world_client.cpp
/*
    实现流程:
        1.包含头文件
        2.初始化 ROS 节点:命名(唯一)
        3.实例化 Action客户端 对象
        4.等待服务端启动完成
        5.实例化目标对象
        6.发送目标任务到服务端
        7.等待任务完成
*/

// 1.包含头文件
#include <ros/ros.h>
#include <sstream>
#include <actionlib/client/simple_action_client.h>
#include <action_hello_world/FindPrimesAction.h>

typedef actionlib::SimpleActionClient<action_hello_world::FindPrimesAction> Client;

// 处理最终结果
void done_cb(const actionlib::SimpleClientGoalState &state, const action_hello_world::FindPrimesResultConstPtr &result)
{
    if (state.state_ == state.SUCCEEDED)
    {
        std::stringstream ss;
        for (const int n : result->primes)
        {
            ss << std::to_string(n) << " ";
        }
        ROS_INFO("%d 以内的质数有: [ %s]", result->number, ss.str().c_str());
    }
}

// 确认目标，服务激活
void active_cb()
{
    ROS_INFO("开始查找...");
}

// 处理连续反馈的定期任务信息
void feedback_cb(const action_hello_world::FindPrimesFeedbackConstPtr &feedback)
{
    ROS_INFO("当前数字: %d, %s质数", feedback->number, feedback->is_prime?"是":"不是");
}

int main(int argc, char **argv)
{
    setlocale(LC_ALL, ""); // 解决中文打印乱码

    // 2.初始化 ROS 节点:命名(唯一)
    ros::init(argc, argv, "action_hello_world_client");
    // 3.实例化 Action客户端 对象
    Client client("/find_primes", true);
    // 4.等待服务端启动完成
    client.waitForServer();
    ROS_INFO("添加任务前的状态: %s", client.getState().toString().c_str());
    // 5.实例化目标对象
    action_hello_world::FindPrimesGoal goal;
    goal.number = 12;
    ROS_INFO("查找 %d 以内的质数", goal.number);
    // 6.发送目标任务到服务端，并注册任务完成、确认目标、定期任务信息的回调函数
    client.sendGoal(goal, &done_cb, &active_cb, &feedback_cb);
    ROS_INFO("添加任务后的状态: %s", client.getState().toString().c_str());

    ros::Duration(1.0).sleep();
    ROS_INFO("执行任务时的状态: %s", client.getState().toString().c_str());
    // 7.等待任务完成
    client.waitForResult(ros::Duration(1000.0)); // 等待结果，1000秒超时
    ROS_INFO("任务执行完的状态: %s", client.getState().toString().c_str());

    return 0;
}